### ISSUE TITLE
EDSM Sync resets system visit totals in database

### ISSUE THREAD

**Issue by rodan123** [EDCD/EDDI#566]
This is an obscure bug I have been chasing for quite some time. I found that EDDI often resets the number of visits to various star systems.

I think I have tracked down the issue. Please confirm.

EDDI syncs at startup with EDSM and collects the number of visits to each star system traveled to since EDDI was last launched.  Date stored in edsm.json "lastSync"

I think the nature of the bug is that EDDI resets the visit data for systems that you have visited between program launches and I think I have tracked the cause to a small error in _StarMapService/StarMapService.cs_ 

Line 257 Reads; 
`CurrentStarSystem.visits = systems[system].visits;`
I believe it should read;
`CurrentStarSystem.visits += systems[system].visits;`

If I am reading correctly, the code sets the number of visits _since the last sync time_ as the total number of visits to the system, instead of adding that value to the stored number of visits to the system, as it should.

<details>
<summary>StarMapService.cs Sync </summary>

public void Sync(DateTime? since = null)
        {
            Logging.Info("Syncing with EDSM");
            try
            {
                Dictionary<string, StarMapLogInfo> systems = getStarMapLog(since);
                Dictionary<string, string> comments = getStarMapComments();
                List<StarSystem> syncSystems = new List<StarSystem>();
                foreach (string system in systems.Keys)
                {
                    StarSystem CurrentStarSystem = StarSystemSqLiteRepository.Instance.GetOrCreateStarSystem(system, false);
                    **CurrentStarSystem.visits += systems[system].visits;**
                    CurrentStarSystem.lastvisit = systems[system].lastVisit;
                    if (comments.ContainsKey(system))
                    {
                        CurrentStarSystem.comment = comments[system];
                    }
                    syncSystems.Add(CurrentStarSystem);
                    if (syncSystems.Count == syncBatchSize)
                    {
                        saveStarSystems(syncSystems);
                        syncSystems.Clear();
                    }
                }
                if (syncSystems.Count > 0)
                {
                    saveStarSystems(syncSystems);
                }
                Logging.Info("EDSM sync completed");
            }
            catch (EDSMException edsme)
            {
                Logging.Debug("EDSM error received: " + edsme.Message);
            }
            catch (ThreadAbortException e)
            {
                Logging.Debug("EDSM update stopped by user: " + e.Message);
            }
        }

</details>

**Comment by Tkael** [comment]
The number of visits returned from EDSM is an absolute value, it's not an incremental value since `lastSync`.
So as an example, according to EDSM my total lifetime visits to Maia is 81 visits. If I constrain the sync period to the last 24 hours, the value returned is still 81 visits. 
Under your proposal, EDDI would add those values together during a sync so the count of visits would become 81 + 81 = 162. I don't think this would be correct.

It's possible that connectivity issues with EDSM may be preventing some visits from being reported to EDSM? 
Also, EDDI will not parse journal entries for events that occur while EDDI is not running.

**Comment by Tkael** [comment]
Interesting. I'll check that out tonight.

**Comment by rodan123** [comment]
Looked at this some more tonight. 
It seems that EDDI alone keeps track of the number of visits to each system and that the EDSM API does not provide that data. [https://www.edsm.net/en/api-logs-v1](https://www.edsm.net/en/api-logs-v1
)
EDDI just counts the number of duplicate entries for each system when you do a full sync with EDSM. EDDI then increments the visits total in the database as you travel to each system.

I am still not clear why EDDI does an incremental sync on startup. But my idea of adding the incremental value doesn't work, because EDDI already accounted for those visits in the database before the incremental sync. And the substitution on the incremental value for the total is surely incorrect. 

I am playing with just skipping updating the count for incremental updates, since it is already correct in the database.

`if (since == null)
                    {
                        CurrentStarSystem.visits = systems[system].visits;
                    }`

This would just change the visited values only when doing a full sync, since there seems to be no need to do a incremental sync to update visit totals. The only downside is that if another program updated EDSM, EDDI wouldn't know about the changes unless you did a full sync.

**Comment by Tkael** [comment]
Ahh, you're right. I didn't look at `getStarMapLog` closely enough. PR submitted with a fix.

**Comment by Tkael** [comment]
The incremental sync updates the database with changes to actual systems (new stations, new factions, etc.) and updates any comments that you may have set on systems,

**Comment by Tkael** [comment]
Please test the PR is can spare a couple of minutes to do so. :-)

**Comment by rodan123** [comment]
@Tkael Sure.. will do.
Question though, looking at the changes;
` if (since != null)
                {
                    /// If we're re-obtaining and resetting the flight logs, we need to replace the value.
                    /// Otherwise, the event handler increments system visits.
                    CurrentStarSystem.visits = systemLogs[system].visits;
                    CurrentStarSystem.lastvisit = systemLogs[system].lastVisit;
                }`

Shouldn't it be  `if (since == null)`?  Because **since** is set to null for a full sync and has a value for a partial sync.

**Comment by Tkael** [comment]
Good catch again. Fix pushed.

**Comment by rodan123** [comment]
The revised code seems to be working well.
I did a full sync from EDSM and travelled in game, noting visit counts.
The full sync replaced the local dates/times I had in the SQL database, with UTC.
Quit and restarted EDDI numerous times. Checked the visit counts after restarting EDDI and all looked good.

@Tkael Great job, thanks for looking at this and for resolving the issue.

*It might be worth mentioning in the release notes that those who used the counters should do a full sync with EDSM to restore the proper visitation counts.
